File: src/main/java/puzzles/year22/day1/Part1.java
package puzzles.year22.day1;

import util.Puzzle;

public class Part1 implements Puzzle<Integer> {
    @Override
    public Integer getAnswer(String inputString) {
        String[] lines = inputString.split(System.lineSeparator());

        int max = 0;
        int currentTotal = 0;

        for (int i = 0; i < lines.length; i++) {
            if (lines[i].isEmpty()) {
                if (currentTotal > max) {
                    max = currentTotal;
                }
                currentTotal = 0;
            } else {
                currentTotal += Integer.parseInt(lines[i]);
            }
        }
        return max;
    }
}

File: src/main/java/puzzles/year24/day19/Part1.kt
package puzzles.year24.day19

import util.Puzzle

class Part1 : Puzzle<Int?> {
    val cache = mutableMapOf<String, Boolean>()

    override fun getAnswer(inputString: String): Int {
        val towels = inputString.split(System.lineSeparator() + System.lineSeparator())[0].split(", ").sortedBy { it.length }.reversed()
        val patterns = inputString.split(System.lineSeparator() + System.lineSeparator())[1].lines()

        var result = 0;
        for (pattern in patterns) {
            if (isPossible(towels, pattern)) {
                result++
            }
        }
        return result
    }

    private fun isPossible(towels: List<String>, pattern: String): Boolean {
        if (pattern.isEmpty()) return true;

        if (cache.containsKey(pattern)) {
            return cache[pattern]!!
        }

        for (towel in towels) {
            if (pattern.startsWith(towel)) {
                if (isPossible(towels, pattern.substring(towel.length))) {
                    cache[pattern] = true
                    return true;
                }
            }
        }

        cache[pattern] = false
        return false;
    }
}
